Design/algorithm_refactoring.md

```diff
@@ -0,0 +1,209 @@
+
+    ```cpp
+    std::bind((double &(std::vector<double>::*)(size_t))&std::vector<double>::operator[], &data, _1)
+    ```
+
+    Note that it is not possible to use implicit type conversion in combination with templates. Suppose we modify `transform` to take a `std::function<T&(size_t)>` as a getter. We then **cannot** use
+
+    ```cpp
+    transform(/*...*/,
+              std::bind((double &(std::vector<double>::*)(size_t))&std::vector<double>::operator[],
+                        &data,
+                       _1), /*...*/);
+     ```
+
+     but need to take a detour via a variable,
+
+    ```cpp
+    auto func = std::bind((double &(std::vector<double>::*)(size_t)) &
+                                      std::vector<double>::operator[],
+                                  &data, _1);
+    transform(/*...*/, func, &data, _1), /*...*/);
+       ```
+
+
+
+Furthermore, we would probably want to provide overloads for supporting the following cases:
+
+- Transform for all spectra.
+- Transform for a range of spectra.
+- Transform for a list of spectra.
+
+
+### Casting of input workspaces to specific workspace types
+
+```cpp
+// Check if its an event workspace
+EventWorkspace_const_sptr eventW =
+    boost::dynamic_pointer_cast<const EventWorkspace>(inputWS);
+if (eventW != NULL) {
+  this->execEvent();
+  return;
+}
+```
+
+A very nice way to solve this would be to use the **visitor pattern**:
+
+- Add the following method to `Algorithm`:
+  ```cpp
+  virtual std::string getInputWorkspaceForVisitorPattern() const {
+    return std::string{};
+  }
+   ```
+  Algorithm implementations that have an input workspace and support the visitor pattern reimplement this function and return the name of the input workspace property.
+- Add a method like this to all Workspace types:
+  ```cpp
+  void Workspace2D::apply(API::Algorithm &algorithm) {
+    algorithm.apply(*this);
+  }
+  ```
+- `Algorithm::Execute()` calls `getInputWorkspaceForVisitorPattern()` to obtain a workspace `ws` and then calls `ws.apply(*this);`
+- All algorithms implement `apply` for various workspace types.
+
+The last bullet is where it all starts to fall apart, there is a fatal flaw in this mechanism: Inheritance does not go well with overloading. For the visitor pattern to work we would need to implement `Algorithm::apply` for **all workspace types**. What we would actually like to do is, e.g., implement `Rebin::apply(const MatrixWorkspace &ws)` and `Rebin::apply(const EventWorkspace &ws)`. Calling the algorithm with a `Workspace2D` will then **not** resolve to `Rebin::apply(const MatrixWorkspace &ws)` but `Algorithm::apply(const Workspace2D &ws)`.
+
+The conclusion is that currently there is **no reasonable way for using the visitor pattern** in this case.
+
+- Are there other ways to solve the some problem?
+- Reconsider the visitor pattern when redesigning workspaces.
+
```

Comment:
If only we had the new re-designed workspaces: http://www.mantidproject.org/DevWorkshop2016_Thu_small_group.